### DISCUSSION TITLE
Way to execute function just before  or after tele-sensor message publish

### DISCUSSION THREAD

**Discussion by usrrsr** [arendst/Tasmota#19583] [General]
Is there any way to call a function just before tasmota publish tele-sensor or tele-state message, so that i can send updated sensor data, 
I tried to call update function from json_append() but this caused partial sensor message string.
Is there any rule event or so to implement 

Regards,

**Comment by barbudor** [comment]
Assuming that you are talking about Berry programming...

json_append() is the where you should add information to the SENSOR message using `tasmota.response_append()`
Without any more details on what code you tried, it is difficult to tell where you missed
You should not spend too much time within json_append()
Exemple is available in the cookook: https://tasmota.github.io/docs/Berry-Cookbook/#creating-an-i2c-driver

**Reply by sfromis** [reply]
If the sensor updating is time-consuming, it may be relevant to have separate ticks to update values at at useful intervals, and then have the `json_append()` and `web_sensor()` callbacks use the last saved values. If your code is doing something like using `tasmota.read_sensors()` to get other sensor values, you should definitely not be doing that from inside the `json_append()` callback.
